### ISSUE TITLE
pprof -topproto generates invalid proto with incomplete symbolization

### ISSUE THREAD

**Issue by vlankhaar** [google/pprof#347]
### What did you do?

```
# perf record --output /tmp/perf.data -- sleep 2
$ pprof --symbolize=none --topproto --output /tmp/profile.pb.gz /tmp/perf.data
$ pprof -top /tmp/profile.pb.gz
```

### What did you expect to see?

```
 Type: cycles_sample
 Showing nodes accounting for 8, 100% of 8 total
       flat  flat%   sum%        cum   cum%
          6 75.00% 75.00%          6 75.00%  [[kernel.kallsyms]]
          1 12.50% 87.50%          1 12.50%  [ld-2.24.so]
          1 12.50%   100%          1 12.50%  [sleep]
```

### What did you see instead?

```
 /tmp/profile.editted.topproto.gz: malformed profile: multiple functions with same id: 1
 pprof: failed to fetch any source profiles
``` 

### Notes

It appears that the -topproto report generator is generating multiple identical functions.

### Input
The input proto [profile.editted.pb.gz](https://github.com/google/pprof/files/1857746/profile.editted.pb.gz)  is lacking function/filename annotations on all locations:

```protobuf
{
  sample_type: {
    type: 1
    unit: 2
  }
  sample: {
    location_id: [ 0x0000000000000001 ]
    value      : [ 6 ]
  }
  sample: {
    location_id: [ 0x0000000000000002 ]
    value      : [ 1 ]
  }
  sample: {
    location_id: [ 0x0000000000000003 ]
    value      : [ 1 ]
  }
  mapping: {
    id      : 0x0000000000000001
    memory_limit     : 0x0000000000000001
  }
  mapping: {
    id      : 0x0000000000000002
    memory_start     : 0xffffffffa3e00000
    memory_limit     : 0xffffffffc021e000
    filename: 4
    build_id: 5
  }
  mapping: {
    id      : 0x0000000000000003
    memory_start     : 0x00007ff9d4239000
    memory_limit     : 0x00007ff9d445f000
    filename: 6
    build_id: 7
  }
  mapping: {
    id      : 0x0000000000000004
    memory_start     : 0x000055c133bb9000
    memory_limit     : 0x000055c133dc2000
    filename: 8
    build_id: 9
  }
  location: {
    id: 0x0000000000000001
    mapping_id: 0x0000000000000002
    address   : 0xffffffffa3e59846
  }
  location: {
    id: 0x0000000000000002
    mapping_id: 0x0000000000000003
    address   : 0x00007ff9d425363f
  }
  location: {
    id: 0x0000000000000003
    mapping_id: 0x0000000000000004
    address   : 0x000055c133bbd6b0
  }
  string_table       : [ "", "cycles_sample", "count", "_unused_", "[kernel.kallsyms]", "3f1f928df57596fce19a4d60531dba081d3e7e29", "/lib/x86_64-linux-gnu/ld-2.24.so", "ef9474f02d7f7a3959b6dae0afe8e6acee1ac5ba", "/bin/sleep", "d4ab980e7d895d7eef7bdeb093648a6a1e080c19" ]
}
```

### Output
The generated topproto 
[profile.editted.topproto.gz](https://github.com/google/pprof/files/1857753/profile.editted.topproto.gz)
 includes line / function information, even though non is provided in the input. 

```protobuf
{       
  sample_type: {        
    type: 1
    unit: 2
  }     
  sample_type: {        
    type: 3
    unit: 2
  }     
  sample: {     
    location_id: [ 0x0000000000000001 ]
    value      : [ 6, 6 ]
  }     
  sample: {     
    location_id: [ 0x0000000000000002 ]
    value      : [ 1, 1 ]
  }     
  sample: {     
    location_id: [ 0x0000000000000003 ]
    value      : [ 1, 1 ]
  }     
  location: {   
    id: 0x0000000000000001
    line: {     
      function_id: 0x0000000000000001
    }   
  }     
  location: {   
    id: 0x0000000000000002
    line: {     
      function_id: 0x0000000000000001
    }   
  }     
  location: {   
    id: 0x0000000000000003
    line: {     
      function_id: 0x0000000000000001
    }   
  }     
  function: {   
    id: 0x0000000000000001
  }     
  function: {   
    id: 0x0000000000000001
  }     
  function: {   
    id: 0x0000000000000001
  }     
  string_table       : [ "", "cum", "count", "flat" ]
  default_sample_type: 0
}
```

**Comment by nolanmar511** [comment]
vlankhaar mentioned that it is probably related to [this](https://github.com/google/pprof/blob/2e397545a68de326eee18011decf2edc2d5f5d7d/internal/report/report.go#L327) line. 

The function is always appended, even when the function may already be in the array.

**Comment by nolanmar511** [comment]
Fixed with #388

**Comment by nolanmar511** [comment]
#388 merged. 
Marking this as fixed.
